### PR TITLE
Use general parameter-shift rules in `param_shift_hessian`

### PR DESCRIPTION
**Pull request by dwierichs**
Here we introduce the general parameter-shift rules to `param_shift_hessian`, similar to #2182 for `param_shift`. As updated in #2227, the `grad_recipe` of an operation is preferred over `parameter_frequencies` or the frequencies derived from `generator`.
For the diagonal of the Hessian, such a custom `grad_recipe` is iterated, reusing code from `generate_shift_rule` with `order=2`.

In addition, some restructuring of helper methods has been performed, hopefully making the module a bit easier to overview.
However, there still is some work to do regarding the structure, which we leave for another PR.

*Related Issues*
#2195

### CODE REVIEW

**Comment by dwierichs** [comment]
[sc-16002]

**Comment by dwierichs** [comment]
@josh146 @dime10 Currently the setup is such that `diagonal_shifts` are expected to be _first order_ shifts, that then are used to create a first-order recipe, which finally is iterated to obtain the second-order recipe for the respective diagonal Hessian entry.
Visually: `diagonal_shifts --> first_order_recipe with diagonal_shifts --iterate--> second_order_recipe`
Users might expect to be able to set the _iterated, second-order_ shifts directly, and it might actually be useful to be able to do so. On the other hand, it might be extra work to provide those second-order shifts if users just want to recycle the `off_diagonal_shifts` for the diagonal as well, and leaving the iteration into a second-order recipe to PL.
What is your opinion here? I see the following options:
1. Leave as is. Users can _not_ set the second-order shifts.
2. Change `diagonal_shifts` to expect the second-order shifts. Users can set the second-order shifts but this may cause extra work if users do not want to provide separate `diagonal_shifts`. Also, setting second-order shifts has the risk of increasing the number of executions, because the default rules (and the result of `general_shift_rules.py::_iterate_shift_rule`) make use of periodicity.
3. Adding a flag indicating which of 1. or 2. is desired by the user.

This might be a topic for a separate PR, and it want to mention that `parameter_shift.py::_get_operation_recipe` currently behaves in accordance with 1. above, which one would need to change if deciding for 2. or 3.

**Comment by josh146** [comment]
@dwierichs while I see the utility in both approaches, I vote for sticking with

> Leave as is. Users can not set the second-order shifts.

for now, and we we always (a) create a feature issue, and (b) re-assess priority if this is requested 🙂

File: pennylane/gradients/param_shift_hessian.py [hunk 1]
```diff
@@ -16,73 +16,105 @@
...
+def expval_hessian_param_shift(
+    tape, argnum, diff_methods, diagonal_shifts, off_diagonal_shifts, f0
+):
     r"""Generate the Hessian tapes that are used in the computation of the second derivative of a
     quantum tape, using analytical parameter-shift rules to do so exactly. Also define a
     post-processing function to combine the results of evaluating the Hessian tapes.
 
     Args:
-        tape (.QuantumTape): input quantum tape
+        tape (pennylane.QNode or .QuantumTape): quantum tape or QNode to differentiate
```
Comment (dime10) [inline pennylane/gradients/param_shift_hessian.py:99]: I don't think this is true. At this point, the input is always a tape I believe.
```suggestion
        tape (.QuantumTape): quantum tape to differentiate
```

File: pennylane/gradients/param_shift_hessian.py [hunk 2]
```diff
@@ -16,73 +16,105 @@
...
+        argnum (int or list[int] or None): Parameter indices to differentiate
+            with respect to. If not provided, the Hessian with respect to all
+            trainable indices is returned. Note that the indices refer to tape
+            parameters if ``tape`` is a tape, and to QNode arguments if it is a QNode.
```
Comment (dime10) [inline pennylane/gradients/param_shift_hessian.py:103]: Similarly, this would hold for the top-level `param_shift_hessian` function, but here they should have been mapped to tape parameters.

File: pennylane/gradients/param_shift_hessian.py [hunk 3]
```diff
@@ -162,29 +193,31 @@ def processing_fn(results):
         out_dim = qml.math.shape(results)[1:]
         # The desired shape of the Hessian is:
         #       (QNode output dimensions, # trainable gate args, # trainable gate args),
-        # but first we accumulate all elements into a list, since no array assingment is possible.
+        # but first we accumulate all elements into a list, since no array assignment is possible.
         hessian = []
         # Keep track of tape results already consumed.
-        start = 1 if unshifted_coeffs and f0 is None else 0
+        start = int(bool(unshifted_coeffs) and f0 is None)
```
Comment (dime10) [inline pennylane/gradients/param_shift_hessian.py:199]: Is this change functionally different? I feel like the previous formulation was slightly more readable :)

File: tests/gradients/test_param_shift_hessian.py [hunk 4]
```diff
@@ -323,6 +415,27 @@ def circuit(x, y, z):
 
         assert all(np.allclose(expected[i], hessian[i]) for i in range(3))
 
+    def test_with_channel(self):
```
Comment (dime10) [inline tests/gradients/test_param_shift_hessian.py:418]: Great to see these additional test scenarios 🎉

**Review by dime10 — CHANGES_REQUESTED** [review]
Amazing work @dwierichs!! 💯 
It's really good to see the Hessian extended with the general parameter shift functionality :)

**Review by josh146 — COMMENTED** [review]
Thought I should submit my accidental code review, back when I thought you had accidentally tagged me! I only got about 2 comments in.

File: pennylane/gradients/__init__.py [hunk 5]
```diff
@@ -282,14 +282,14 @@ def my_custom_gradient(tape, **kwargs):
 """
 import pennylane as qml
 
-from . import finite_difference
 from . import parameter_shift
 from . import parameter_shift_cv
 from . import param_shift_hessian
+from . import finite_difference
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/__init__.py:288]: I guess not, it just feels like the parameter shift pipelines are the predominant part of the module :) :D

File: pennylane/gradients/general_shift_rules.py [hunk 6]
```diff
@@ -52,16 +93,16 @@ def process_shifts(rule, tol=1e-10, check_duplicates=True):
     rule = rule[:, ~(rule[0] == 0)]
 
     if check_duplicates:
+        if rule.shape[0] == 3:
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/general_shift_rules.py:96]: Yes, one definitely could include them in the same. I went for the split here to affect the scenarios without multipliers as little as possible. In general, there should be some clean up work done in a separate PR regarding grad_recipe formatting, processing and where the multipliers should be considered - and where not.
Regarding the docstring, that's simply outdated, thanks! :)

File: pennylane/gradients/general_shift_rules.py [hunk 7]
```diff
@@ -314,11 +418,75 @@ def generate_multi_shift_rule(frequencies, shifts=None, orders=None):
...
+
+def generate_multishifted_tapes(tape, idx, shifts, mults):
+    r"""Generate a list of tapes where the corresponding trainable parameter
+    indices have been shifted by the values given.
+
+    Args:
+        tape (.QuantumTape): input quantum tape
+        idx (list[int]): trainable parameter indices to shift the parameters of
+        shifts (list[list[float or int]]): nested list of shift values, each
+            list containing a value for each index
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/general_shift_rules.py:467]: Good catch - and great suggestion regarding general input types and explainers :)

File: pennylane/gradients/general_shift_rules.py [hunk 8]
```diff
@@ -314,11 +418,75 @@ def generate_multi_shift_rule(frequencies, shifts=None, orders=None):
...
+        mults: TODO
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/general_shift_rules.py:468]: Oops :see_no_evil: Thanks :)

File: pennylane/gradients/general_shift_rules.py [hunk 9]
```diff
@@ -314,11 +418,75 @@ def generate_multi_shift_rule(frequencies, shifts=None, orders=None):
...
+
+    Returns:
+        list[QuantumTape]: List of quantum tapes. Each tape has multiple parameters
+            (indicated by ``idx``) shifted by the values of ``shifts``. The length
+            of the returned list of tapes will match the length of ``shifts``.
+    """
+    params = list(tape.get_parameters())
+    mults = [None] * len(shifts) if mults is None else mults
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/general_shift_rules.py:476]: Ah yes, we actually know that `shifts` is a non-ragged array! Very nice :)

File: pennylane/gradients/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/__init__.py:288]: Ah and regarding the order of import, it is about cyclic imports that were caused by moving around helper functions :)

File: pennylane/gradients/param_shift_hessian.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/gradients/param_shift_hessian.py:103]: Oh, yes, you're right of course (also with the above).

File: pennylane/gradients/param_shift_hessian.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/gradients/param_shift_hessian.py:199]: Mhm. No it is not, and I actually agree it was more readable before... This might be a leftover from an attempt to restructure the tape list a little, which I didn't end up keeping. Will revert :)

File: pennylane/gradients/parameter_shift.py [hunk 10]
```diff
@@ -69,11 +74,18 @@ def _square_observable(obs):
...
 
 
-def _get_operation_recipe(tape, t_idx, shifts):
+def _get_operation_recipe(tape, t_idx, shifts, order=1):
     """Utility function to return the parameter-shift rule
     of the operation corresponding to trainable parameter
     t_idx on tape.
 
+    Args:
+        tape (.tape.QuantumTape): Tape containing the operation to differentiate
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift.py:83]: Me neither :sweat_smile: I thought because `qml.QuantumTape()` does not work, one needs the extra `.tape`...

File: pennylane/gradients/parameter_shift.py [hunk 11]
```diff
@@ -592,7 +629,7 @@ def param_shift(
             "If this is unintended, please mark trainable parameters in accordance with the "
             "chosen auto differentiation framework, or via the 'tape.trainable_params' property."
         )
-        return [], lambda _: ()
+        return [], lambda _: np.zeros((tape.output_dim, 0))
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift.py:632]: If I'm not mistaken, `param_shift` is a somewhat more low-level function than `qml.grad` and `qml.jacobian`. For example, the derivatives for parameters that are not included in `argnum`, if given, are set to `0` instead of being skipped by `param_shift`, whereas `qml.grad` would not return any object/entry for such parameters...

File: tests/gradients/test_general_shift_rules.py [hunk 12]
```diff
@@ -16,14 +16,66 @@
 
 import numpy as np
 import pennylane as qml
-from pennylane import numpy as np
+
+# from pennylane import numpy as np
```
[the remark this answers is not in the document]
Comment (dwierichs) [reply tests/gradients/test_general_shift_rules.py:20]: Thanks :)

**Comment by dwierichs** [comment]
Thanks @dime10  and @josh146 !
I'll take care of the open points next week :)

File: pennylane/gradients/general_shift_rules.py [hunk 13]
```diff
@@ -314,11 +418,75 @@ def generate_multi_shift_rule(frequencies, shifts=None, orders=None):
...
-    combined_rules = []
+    return _combine_shift_rules(rules)
 
-    for partial_rules in itertools.product(*rules):
-        c, s = np.stack(partial_rules).T
-        combined = np.concatenate([[np.prod(c)], s])
-        combined_rules.append(np.stack(combined))
 
-    return np.stack(combined_rules).T
+def generate_shifted_tapes(tape, idx, shifts, multipliers=None):
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/gradients/general_shift_rules.py:424]: Might make sense to merge it.

The only argument I have against merging it would be if this simpler case is written to be more efficient

File: pennylane/gradients/general_shift_rules.py [hunk 14]
```diff
@@ -314,11 +418,78 @@ def generate_multi_shift_rule(frequencies, shifts=None, orders=None):
...
+        shifted_tape.set_parameters(new_params)
+        tapes.append(shifted_tape)
+
+    return tapes
+
+
+def generate_multishifted_tapes(tape, indices, shifts, multipliers=None):
+
+    r"""Generate a list of tapes where the corresponding trainable parameter
+    indices have been shifted by the values given.
```
Comment (josh146) [inline pennylane/gradients/general_shift_rules.py:462]: This docstring summary is identical to the function above :)

File: pennylane/gradients/general_shift_rules.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/general_shift_rules.py:96]: I took another look and found that merging the two functionalities actually is easier than I thought, so I did it :)

File: pennylane/gradients/general_shift_rules.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/general_shift_rules.py:424]: I tried to merge them but there is an issue arising from the `dtype = new_params[idx].dtype` line that is making problems for iterated differentiation, which I don't really understand. I'll leave this to the clean up PR.

File: pennylane/gradients/general_shift_rules.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dime10) [reply pennylane/gradients/general_shift_rules.py]: Oh that's great to hear :)

File: pennylane/gradients/parameter_shift.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dime10) [reply pennylane/gradients/parameter_shift.py]: Ah thanks for explaining this Josh :)

File: pennylane/gradients/parameter_shift.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dime10) [reply pennylane/gradients/parameter_shift.py:632]: Ah I see now. Well `argnum` behaves a bit strangely for the gradient transforms, as it only allows to select/deselect arguments that are already in the `tape.trainable_params`. But the "true" number of trainable tape parameters does indeed directly determine the last dimension of the output gradient array, so returning an empty array when there are no trainable parameters is the natural extension here 👍 
In this case, would mind changing this line for the other gradient transforms as well (finite-diff, cv)? 

Although practically, I'm not sure the dimension information in empty return array with some is of any use. Do you know of some? :) There is also the problem that `tape.output_dim` only contains the "flattened" tape output dimensions (e.g. it will not give the correct dimensions if the tape contains `qml.probs(0)` and `qml.probs(1)`, a `(2, 2)` output shape).

**Review by dime10 — APPROVED** [review]
Fantastic stuff @dwierichs 🎉 🎉

File: pennylane/gradients/param_shift_hessian.py [hunk 15]
```diff
@@ -211,17 +241,24 @@ def param_shift_hessian(tape, f0=None):
...
-        "RX", "RY", "RZ", "Rot", "PhaseShift", "ControlledPhaseShift", "MultiRZ", "PauliRot",
-        "U1", "U2", "U3", "SingleExcitationMinus", "SingleExcitationPlus", "DoubleExcitationMinus",
-        "DoubleExcitationPlus", "OrbitalRotation".
-
     Args:
         tape (pennylane.QNode or .QuantumTape): quantum tape or QNode to differentiate
+        argnum (int or list[int] or None): Parameter indices to differentiate
+            with respect to. If not provided, the Hessian with respect to all
+            trainable indices is returned. Note that the indices refer to tape
+            parameters if ``tape`` is a tape, and to QNode arguments if it is a QNode.
```
Comment (dime10) [inline pennylane/gradients/param_shift_hessian.py:249]: Actually I think the `argnum` indices always refer to the trainable tape arguments even for qnodes, see this example:

```py
@qml.qnode(qml.device("default.qubit", wires=2))
def circuit(x):
    qml.RY(x[0], wires=0)
    qml.RX(x[1], wires=0)
    return qml.probs(0), qml.probs(1)
qml.gradients.param_shift(circuit, argnum=1)(qml.numpy.array([0.1, 0.2]))
```

File: pennylane/gradients/parameter_shift.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dwierichs) [reply pennylane/gradients/parameter_shift.py:632]: >Ah I see now. Well argnum behaves a bit strangely for the gradient transforms, as it only allows to select/deselect arguments that are already in the tape.trainable_params.

Yes, I feel like we really should work a bit more towards a consistent system, also regarding wording, across the codebase (also referring to our chat about `argnum`). 

> In this case, would you mind changing this line for the other gradient transforms as well (finite-diff, cv)?

Done :)

>Although practically, I'm not sure the dimension information in the empty return array is of any use. Do you know of some? 

That is a fair point. I am not sure, but I personally thought of homogeneous return types being more convenient in general. Something like `hessian.shape` would work even if there are no trainable parameters, for example :)

>There is also the problem that tape.output_dim only contains the "flattened" tape output dimensions

That's a good point! :thinking: I suppose for now we may stick with this though? Is there a plan to adapt `tape.output_dim` to such non-flat cases?

File: pennylane/gradients/param_shift_hessian.py (hunk 15 above)
Comment (dwierichs) [reply pennylane/gradients/param_shift_hessian.py:249]: Adapted the docstring, all further edits regarding the logic should be taken care of together with `param_shift`

File: pennylane/gradients/param_shift_hessian.py [hunk 16]
```diff
@@ -276,12 +276,6 @@ def param_shift_hessian(tape, argnum=None, diagonal_shifts=None, off_diagonal_sh
         - If the input is a tape, a tuple containing the list of parameter-shifted tapes, and a
           post-processing function to be applied to the evaluated tapes, is returned.
 
-        Note: By default a QNode with the keyword ``hybrid=True`` computes derivates with respect to
-        QNode arguments, which can include classical computations on those arguments before they are
-        passed to quantum operations. The "purely quantum" Hessian can instead be obtained with
-        ``hybrid=False``, which is then computed with respect to the gate arguments and produces a
-        result of shape ``(*QNode output dimensions, # gate arguments, # gate arguments)``.
-
```
Comment (dime10) [inline pennylane/gradients/param_shift_hessian.py:247]: Why was this removed? :)

File: pennylane/gradients/parameter_shift.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dime10) [reply pennylane/gradients/parameter_shift.py:632]: >That's a good point! 🤔 I suppose for now we may stick with this though? Is there a plan to adapt tape.output_dim to such non-flat cases?

Yes I believe this PR is working on it actually https://github.com/PennyLaneAI/pennylane/pull/2044 :)

File: pennylane/gradients/param_shift_hessian.py (hunk 16 above)
Comment (dwierichs) [reply pennylane/gradients/param_shift_hessian.py:247]: oh dear, I did not focus properly there, and thought it was carried over from a different file without being applicable, but of course it is. Thanks! :)
